### PR TITLE
Add support for commands nesting, remove commands list from robot, clean-up notifications

### PR DESCRIPTION
**Pull request by astaff**
## Add support for commands nesting

`commands` attribute in `Session` now reflects nesting levels for protocol commands. This is achieved by session listening to `robot.command` and adding it to the stack on *before* event and removing it on *after*. This allows to track stack depth for each command being executed. Upon receiving all the commands during simulation we can reconstruct the tree through DFS traversal.

## Remove commands list from robot

As an effort to remove state from `robot`, made `robot` and `pipette` emit `robot.command` and made `session` responsible for handling `robot.command`, mutating itself and emitting`session.state.change` sending `session` value that can be consumed by the UI.

Removal of commands, affected a lot of pipette tests that had to be refactored. To reduce code re-write, test setup monkeypatches `robot.commands` using session.

## Clean-up notifications

As an effort to keep session code clean, following conversation with @jaredgreene1 and @mcous , made the following changes to EventBroker:
  - Removed it from `utils` and moved to separate module, `opentrons.broker`.
  - Moved functions from `EventBroker` class into `broker` module, as a cleaner way to implement singleton.
  - Allowed listeners to `subscribe` to select names. With `handler` defined, it will call a handler passing event `name` and `payload`. If handler is not defined return `async iterable` to be used in `async for` statements. 
  - Moved `Notifications` from `session` to `broker`.
  - Removed `@traceable` and introduced `publish.before, publish.after and publish.both` allowing event to be emitted before function call, after function call or both. 
  - Made decorator signature more explicit by removing name inference from function's FQN and adding a named argument.
  - Made `broker.notify` take `name` as an argument
  - Refactored code that used EventBroker

## Notes

- There *are* breaking changes in the API that need to be further explained: `robot.commands` is gone. We need to implement a protocol developer friendly way to use `session` for running the protocol and retrieving commands.
- Per @jaredgreene1 and @mcous conversation, we should try to make event emission more explicit and have an option to define message types and notification factory, that would produce `notify` functions with named arguments.

Check List:

- [x] Are there breaking changes in the API and how are they being communicated?
- [ ] Who is reviewing your code?

### CODE REVIEW

File: api/opentrons/commands/types.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+MAGBEAD_ENGAGE = makeRobotCommandName('MAGBEAD_ENGAGE')
+DELAY = makeRobotCommandName('DELAY')
+BLOW_OUT = makeRobotCommandName('BLOW_OUT')
+AIR_GAP = makeRobotCommandName('AIR_GAP')
+TOUCH_TIP = makeRobotCommandName('TOUCH_TIP')
+RETURN_TIP = makeRobotCommandName('RETURN_TIP')
+HOME = makeRobotCommandName('HOME')
+
+# State #
+STATE_CHANGE = 'session.state.change'
```
Comment (mcous) [inline api/opentrons/commands/types.py]: I don't think this is related to commands. Why is this here? Also, it doesn't appear to be used

File: api/opentrons/session/session.py [hunk 2]
```diff
@@ -1,37 +1,44 @@
...
 class SessionManager(object):
-    def __init__(self, loop=None, filters=[
-            'add-command',
-            'session.state.change']):
-        self.notifications = Notifications(loop=loop, filters=filters)
+    def __init__(self, loop=None):
+        self.unsubscribe, self.notifications = \
+            subscribe('session.state.change', loop=loop)
         self.robot = Robot()
+        self.sessions = []
```
Comment (mcous) [inline api/opentrons/session/session.py]: I know we've talked about multiple sessions at a high level but I don't think there was enough information in place to implement anything. I'm not requesting this be reverted, because whatever, but heads up.

File: api/opentrons/session/session.py [hunk 3]
```diff
@@ -1,37 +1,44 @@
...
+    def clear(self):
+        for session in self.sessions:
+            session.close()
+        self.sessions.clear()
 
     def create(self, name, text):
+        self.clear()
+
         with self.notifications.snooze():
             self.session = Session(name=name, text=text)
```
Comment (mcous) [inline api/opentrons/session/session.py]: Generally, would it be ok to defined all future properties in the constructor with some default value or `None`? I know it's Python and you don't _need_ to do this (same story for JS), but it makes the classes way more readable to know right away  what properties could be present without having to dig through the methods

File: api/opentrons/broker/broker.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+from contextlib import contextmanager
+
+listeners = []
+
+
+class Notifications(object):
+    def __init__(self, loop=None):
+        self.loop = loop or asyncio.get_event_loop()
+        self.queue = Queue(loop=self.loop)
+        self.snoozed = False
```
Comment (mcous) [inline api/opentrons/broker/broker.py]: What's the value of snooze over simply unsubscribing and resubscribing?

File: api/opentrons/session/session.py [hunk 5]
```diff
@@ -119,31 +169,50 @@ def load_commands(self, commands):
         updates self.commands with a dictionary that holds
         a command tree.
         """
-        def children(commands, level=0, base_index=0):
+        def subtrees(commands, level):
```
Comment (mcous) [inline api/opentrons/session/session.py]: These methods seem to me like they should live in commands and not session. Any thoughts?

File: api/opentrons/session/session.py [hunk 6]
```diff
@@ -119,31 +169,50 @@ def load_commands(self, commands):
...
+            {
                 'timestamp': datetime.utcnow().isoformat(),
                 'error': error
             }
-        })
+        )
+        self.on_state_changed()
+
+    def on_state_changed(self):
+        notify('session.state.change', copy.deepcopy(self))
```
Comment (mcous) [inline api/opentrons/session/session.py]: `session.state.change` is a lot more information than we need, especially considering that we're always sending the whole session. Could we perhaps just use `session` and reserve more detailed notifications for smaller diffs if and when we do that?

File: api/tests/opentrons/protocol/test_command_tree.py [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# from opentrons.session import Session
+
+
+# def test_command_tree(protocol):
```
Comment (mcous) [inline api/tests/opentrons/protocol/test_command_tree.py]: Is our command tree stuff uncovered at the moment?

File: api/opentrons/session/session.py (hunk 5 above)
Comment (astaff) [reply api/opentrons/session/session.py]: 👍

File: api/opentrons/session/session.py (hunk 6 above)
Comment (astaff) [reply api/opentrons/session/session.py]: Will do

File: api/tests/opentrons/protocol/test_command_tree.py (hunk 7 above)
Comment (astaff) [reply api/tests/opentrons/protocol/test_command_tree.py]: covered in session/test_session.py. Renaming the function to be more descriptive

File: api/opentrons/broker/__init__.py [hunk 8]
```diff
@@ -0,0 +1,3 @@
+from .broker import subscribe, notify
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/broker/__init__.py]: I looked it up and it seems like the way to go is to have all the modules in the root of the package directory.  (https://docs.python.org/3/tutorial/modules.html#packages)

File: api/opentrons/session/session.py (hunk 2 above)
Comment (astaff) [reply api/opentrons/session/session.py]: ok

File: api/opentrons/broker/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/broker/__init__.py]: Our directory structure *does* need some love. But that should be a different PR dedicated to that.

File: api/opentrons/commands/types.py (hunk 1 above)
Comment (astaff) [reply api/opentrons/commands/types.py]: 👍

File: api/opentrons/session/session.py (hunk 3 above)
Comment (astaff) [reply api/opentrons/session/session.py]: I fully agree with that and looking for that in code reviews. Missed this one. Thanks.

File: api/opentrons/broker/broker.py (hunk 4 above)
Comment (astaff) [reply api/opentrons/broker/broker.py]: `self.notifications` is being looped through by the server. Re-subscribing will change that reference.

File: api/tests/opentrons/protocol/test_command_tree.py (hunk 7 above)
Comment (astaff) [reply api/tests/opentrons/protocol/test_command_tree.py]: Ended up moving it to commands/test_tree.py

File: api/opentrons/commands/commands.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
+from . import types
+from ..broker import notify
+import functools
+import inspect
+
+
+def make_command(name, payload):
+    return (name, payload)
```
Comment (mcous) [inline api/opentrons/commands/commands.py]: Why use a tuple here rather than an object? Just curious

File: api/opentrons/commands/commands.py [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+
+
+def home(axis):
+    text = 'Homing pipette plunger on axis {axis}'.format(axis=axis)
+    return make_command(
+        name=types.HOME,
+        payload={
+            'axis': axis,
+            'text': text
```
Comment (mcous) [inline api/opentrons/commands/commands.py]: Do we think `text` should be part of the payload? No need to change, but personally I feel like it's meta information (maybe we could mirror FSA and have a meta key next to name and payload)

File: api/opentrons/commands/commands.py [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+def magbead():
+    pass
+
+
+magbead.engage = engage
+magbead.disengage = disengage
+magbead.delay = delay
+
+
+def publish(before, after, name=None, command=None, payload=None):
```
Comment (mcous) [inline api/opentrons/commands/commands.py]: I don't see `name` used anywhere in source. Is it still needed?

File: api/opentrons/robot/robot.py [hunk 12]
```diff
@@ -798,17 +768,6 @@ def diagnostics(self):
...
-
-        Returns
-        -------
-        A list of string values for each command in the queue, for example:
-
-        ``'Aspirating 200uL at <Deck>/<Slot A1>/<Container plate>/<Well A1>'``
-        """
-        return self._commands
-
+    @commands.publish.before(command=commands.comment)
```
Comment (mcous) [inline api/opentrons/robot/robot.py]: Does `robot.comment` seem like something we should deprecate?

File: api/opentrons/broker/broker.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        listeners.remove(listener)
+
+    if notifications:
+        return (unsubscribe, notifications)
+    return (unsubscribe,)
+
+
+def notify(name, payload):
+    for prefix, listener in listeners:
+        if name.startswith(prefix):
```
Comment (mcous) [inline api/opentrons/broker/broker.py]: I feel like we talked about starting with simple, top-level "topics" (names, whatever) with more specific types in the payload itself, rather than doing any match stuff on the command name. More like:

```py
def handle_command (command):
  if command.type === commands.types.ASPIRATE:
    # aspirate specific stuff
  # ...

subcribe('command', handle_command)
notify('command', {'name': commands.types.ASPIRATE, 'payload': {'foo': 'bar'}})
```

File: api/opentrons/commands/types.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+ROBOT_COMMAND = 'robot.command'
+
+
+# Robot #
+def makeRobotCommandName(name):
+    return '{}.{}'.format(ROBOT_COMMAND, name)
```
Comment (mcous) [inline api/opentrons/commands/types.py]: I noticed that the `robot.command` part of these names is being used in the notification logic (see my comment in `broker.py`). Just wanted to point out that similar code in the front-end (prefixing action names in a given state module) is purely used for collision avoidance and human readability

File: api/opentrons/session/session.py [hunk 15]
```diff
@@ -87,63 +143,56 @@ def run(self, devicename=None):
...
+        return {
+            'name': self.name,
+            'state': self.state,
+            'protocol_text': self.protocol_text,
+            'commands': self.commands.copy(),
+            'command_log': self.command_log.copy(),
+            'errors': self.errors.copy()
+        }
+
+    def on_state_changed(self):
```
Comment (mcous) [inline api/opentrons/session/session.py]: Can we make this method "private"?

File: api/tests/opentrons/conftest.py [hunk 16]
```diff
@@ -103,3 +104,30 @@ def finalizer():
...
+    assert len(result) == len(expected_re), \
+        'result and expected have different length'
+
+    for res, exp in zip(result, expected_re):
+        assert re.compile(
+            exp.lower()).match(res.lower()), "{} didn't match {}" \
+            .format(res, exp)
+
+
+def patch_robot(robot, commands):
```
Comment (mcous) [inline api/tests/opentrons/conftest.py]: Why are we doing this? If it's to avoid having to change tests to use the updated robot API, can we add a TODO?

File: api/opentrons/commands/commands.py (hunk 11 above)
Comment (astaff) [reply api/opentrons/commands/commands.py]: Agreed.

File: api/opentrons/broker/broker.py (hunk 13 above)
Comment (astaff) [reply api/opentrons/broker/broker.py]: We did. I don't remember arriving to any conclusion though. I am leaning towards using name-spaced names and filter by prefix which. Do you have a compelling use case for topics and names in mind?

File: api/opentrons/commands/commands.py (hunk 9 above)
Comment (astaff) [reply api/opentrons/commands/commands.py]: Objects in Python are hard to construct inline. You either have to use `namedtuple` or construct some dummy object and assign it's attributes.  I usually use tuples if there are 2-3 values to unpack. Another option is to use `dict`, but that's string keys. I like `ECMAScript` approach to dicts / objects better.

File: api/opentrons/commands/commands.py (hunk 10 above)
Comment (astaff) [reply api/opentrons/commands/commands.py]: I feel adding the next round of changes will reveal the answer

File: api/opentrons/commands/commands.py [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+
+    # Update / insert values for positional args
+    res.update(dict(zip(inspect.getargspec(f).args, args)))
+
+    # Update it with values for named args
+    res.update(kwargs)
+    return res
+
+
+publish.before = functools.partial(publish, before=True, after=False)
```
Comment (astaff) [inline api/opentrons/commands/commands.py]: @mcous here's an example of in-line construction. publish is a function which is an object, so we are adding attributes, while still being able to call publish, and having easy to read `publish.before` constructs.

File: api/opentrons/commands/types.py (hunk 14 above)
Comment (astaff) [reply api/opentrons/commands/types.py]: I guess this one is the example of where topics could be considered. Do you see a problem with the approach I am taking?

File: api/opentrons/robot/robot.py (hunk 12 above)
Comment (astaff) [reply api/opentrons/robot/robot.py]: It is used to emit a message in command log. Don't see a reason to deprecate it.

File: api/opentrons/commands/commands.py (hunk 9 above)
Comment (mcous) [reply api/opentrons/commands/commands.py]: Sorry, I should've been more specific. By object I did mean dict

File: api/opentrons/broker/broker.py (hunk 13 above)
Comment (mcous) [reply api/opentrons/broker/broker.py]: It's more of a start simpler before getting more complex sort of situation for me. At the moment, we're not doing listening for anything more complex than `session` or `robot.command`, so if we can get away with a simpler interface, then we should because it's less surface area

File: api/opentrons/commands/types.py (hunk 14 above)
Comment (mcous) [reply api/opentrons/commands/types.py]: Not any real problems, no, just that it wasn't immediately obvious to me (until I had gone through the review and then re-read the implementation of `notify`) that listening for, say, `robot` will implicitly return to you all events matching `robot*`. If we're going to match, I feel like we should be explicit about it, and right now, I don't see any need for matching.

Might lead to some weird bugs if we accidentally name one event a string that is the beginning of another unrelated string... but that may or may not be likely

File: api/opentrons/broker/broker.py (hunk 13 above)
Comment (mcous) [reply api/opentrons/broker/broker.py]: See my comment below

File: api/opentrons/broker/broker.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
+import asyncio
+
+from asyncio import Queue
+from concurrent import futures
+from contextlib import contextmanager
+
+subscriptions = {}
```
Comment (mcous) [inline api/opentrons/broker/broker.py]: I'm not in love with this. Have we considered doing something where a single instance of Broker (with notify and subscribe) is created at boot and then passed where it needs to be?

File: api/opentrons/session/session.py [hunk 19]
```diff
@@ -1,37 +1,48 @@
...
-    def __init__(self, loop=None, filters=[
-            'add-command',
-            'session.state.change']):
-        self.notifications = Notifications(loop=loop, filters=filters)
+    def __init__(self, loop=None):
+        self.unsubscribe, self.notifications = \
+            subscribe(SESSION_TOPIC, loop=loop)
+        self.session = None
         self.robot = Robot()
+        self.sessions = []
```
Comment (jaredgreene1) [inline api/opentrons/session/session.py]: Why would we have multiple sessions?

File: api/tests/opentrons/broker/test_broker.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+def my_command(arg1, meta=None, arg2='', arg3=''):
+    return {
+        'name': 'command',
+        'payload': {
+            'description': meta.format(arg1=arg1, arg2=arg2, arg3=arg3)
+        }
+    }
+
+
+@commands.publish.both(command=my_command, meta='{arg1} {arg2} {arg3}')
```
Comment (jaredgreene1) [inline api/tests/opentrons/broker/test_broker.py]: I don't see why we need a decorator for this

File: api/opentrons/broker/broker.py [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+            futures.wait([future])
+
+    async def __anext__(self):
+        return await self.queue.get()
+
+    def __aiter__(self):
+        return self
+
+
+def subscribe(topic, handler=None, loop=None):
```
[the remark this answers is not in the document]
Comment (mcous) [reply api/opentrons/broker/broker.py]: I think that's the long term goal. For now I'm personally ok with this interface until we can have everything (async and sync) use the explicit handler interface

File: api/opentrons/commands/commands.py [hunk 22]
```diff
@@ -0,0 +1,292 @@
+from . import types
+from ..broker import notify
```
[the remark this answers is not in the document]
Comment (mcous) [reply api/opentrons/commands/commands.py]: I kinda prefer relative imports, but that's because I'm a dirty javascript developer

File: api/opentrons/session/session.py (hunk 19 above)
Comment (mcous) [reply api/opentrons/session/session.py]: I made [this same comment](https://github.com/OpenTrons/opentrons/pull/336#discussion_r139224735). We've talked at a high-level about the concept of archived sessions and the need some interface for retrieving them. I think the inclusion of this property is a little premature, but it'll likely be reworked during our larger data layer work

File: api/tests/opentrons/broker/test_broker.py (hunk 20 above)
Comment (mcous) [reply api/tests/opentrons/broker/test_broker.py]: The both publish is used by the command log to properly nest child commands (basically, the message the is sent before adds a parent to the stack, and the after message pops it, I think)

Comment (jaredgreene1) [reply api/tests/opentrons/broker/test_broker.py]: Maybe we can change the naming of this then? Would something like `logged_command` be clearer?

File: api/opentrons/robot/robot.py (hunk 12 above)
Comment (jaredgreene1) [reply api/opentrons/robot/robot.py]: I think `robot.print` would be clearer

File: api/opentrons/broker/broker.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (jaredgreene1) [reply api/opentrons/broker/broker.py]: Why can't we just check for a keyword rather than `None`? Can we just say `async=True` or something like that?

File: api/opentrons/session/session.py (hunk 19 above)
Comment (jaredgreene1) [reply api/opentrons/session/session.py]: Spoke with Mike - I think this should at least be noted by a comment. Otherwise, this is tribal knowledge

File: api/opentrons/broker/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jaredgreene1) [reply api/opentrons/broker/__init__.py]: Yeah, I had some CI breakage because I had left out `__init__.py`

File: api/opentrons/robot/robot.py (hunk 12 above)
Comment (astaff) [reply api/opentrons/robot/robot.py]: I think it's out of scope of this PR, pls open a GitHub issue

File: api/opentrons/broker/broker.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/broker/broker.py]: Removed notifications from subscribe altogether, doing composition in SessionManager for now. Made handler explicit.

File: api/tests/opentrons/broker/test_broker.py (hunk 20 above)
Comment (astaff) [reply api/tests/opentrons/broker/test_broker.py]: Not sure I understand why specifics of how this is being used in one use case, should be reflected in the name.

File: api/opentrons/session/session.py (hunk 19 above)
Comment (astaff) [reply api/opentrons/session/session.py]: Comment added
